DOCS-7779 4.3 DW Type System-kt (#1326)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/dataweave-type-system.adoc

= Type System
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> DataWeave 2.0 introduced a new type system on the language. Learn how to describe constraint expressions for variables and functions to take advantage of the type checking that the type system executes. To understand how values for these types are defined see xref:dataweave-types.adoc[Value Constructs for Types] documentation.

Reviewer: Not sure we need to say what's new in regular docs.

> DataWeave 2.0 introduced a new type system on the language. Learn how to describe constraint expressions for variables and functions to take advantage of the type checking that the type system executes. To understand how values for these types are defined see xref:dataweave-types.adoc[Value Constructs for Types] documentation.

Reviewer: ```suggestion
DataWeave 2.0 supports a type system. To take advantage of the type checking that the type system executes, you need to provide constraint expressions for variables and functions described in this section. For information on how values for these types are defined, refer to xref:dataweave-types.adoc[Value Constructs for Types].
```

The types supported by the DataWeave type system are divided into three categories:

* Simple Types
* Composite Types
* Complex Types

== Simple Types

Simple types represent values such as string, Booleans, etc. These values are atomic, they are not composed by other values. These are the simple types:

* String
* Boolean
* Number
* Regex
* Null
* Temporal: Date, DateTime, LocalDateTime, LocalTime, Time, Period
//Key, KeyValuePairType, NameType, NameValuePairType, NamespaceType, RangeType , ReferenceType, TypeParameter, TypeType, UriType

=== Null

Null is a type of only one value, the value `null`, which means that this value `null` cannot be assigned to the type String or any other type except for the type Null.

For example, the repeat function in the String Module has the signature `repeat(String, Number): String`, which means that it only accepts as a first parameter a value of type String, and as a second parameter a value of type Number. The function returns a value of type String.

> ----

Reviewer: ```suggestion
.DataWeave Script:
----
%dw 2.0
import * from dw::core::Strings
output application/json
```
Not sure what output format or MIME type we want to use since we're not showing the actual output data, but we need to provide output and import directives to avoid throwing an error in Studio.

Also, I'm actually seeing `.Transform` rather than `.DataWeave Script:` elsewhere in this file, but I think the latter might be even more clear than the former throughout this article.

Reviewer: I'd do something like the above on all the examples that import a String function. 

You'll need to modify the import directive to match other function modules that you import, which is essentially all functions except for the Core and Mule functions, for example, `import * from dw::core::Arrays`. Check here for examples for those modules: https://docs.mulesoft.com/mule-runtime/4.2/dw-functions#dw_modules.

Author: Updated.
@afelisatti Could you review the samples and check the function modules for `import * from dw::core::` I was not sure about some of these. So you might leave a comment to have it updated. Thank you.

= Type System
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

DataWeave 2.0 introduced a new type system on the language. To understand how values for these types are defined see xref:dataweave-types.adoc[Value Constructs for DataWeave Types] documentation.

A type system defines a set of constraints to a set of constructs such as:

* Variables
* Function parameters

These constraints are used in the type checking phase where DataWeave ensures that the values assigned
to variables or the arguments for a function call respects its constraints.

For example:

----
var userName: String = "John"
----

The variable definition for `userName` has the constraint of assigning a value of type String.

> If some other type value is assigned, a type checking error raises.

Reviewer 2: `raises` -> `is raised`?

Author: Updated.

----
fun toUser(id: Number, userName: String): String = "you called the function toUser!"
----

> The constraints to call the function `toUser` are that the first argument has to be of type Number, the second one of type String (or to be coercible to those types) and the result of the function has to be of type String.

Reviewer 3: ```suggestion
The constraints to call the function `toUser` are that the first argument has to be of type Number, the second one of type String (or to be coercible to those types). Another constraint here is that the result of the function has to be of type String (this constraint isn't used in the call for `toUser`).
```

Author: Updated.

----
toUser(123, "John")
toUser("123", true)
----

> Both calls don't throw an exception. In the case of the second function call, there is an autocoercion of the String `"123"` to the Number value `123`, and the Boolean `true` to the String value `"true"` to being able to call the function. +

Reviewer 2: Here I think we need something more like: `Though you would expect the second call to fail autocoercions allow it to be successful`

Author: @afelisatti Would the above sentence replace the entire paragraph then? Meaning that this is not necessary: _the String `"123"` to the Number value `123`, and the Boolean `true` to the String value `"true"`_

Reviewer 2: No, this would be more like an addition. Something like `Though you would expect the second call to fail, autocoercions allow it to be successful and both of them work. In the case of the second function call, there is an autocoercion of the String ...`

Author: Updated

An example of a call that throws a type checking error is `toUser("a 12", "John")` because the String `"a 12"` isn't coercible to a Number. +

> DataWeave also uses a global type inference algorithm to validate your code even if no type is specified as

Reviewer 2: `type is` -> `types are`?

Author: Updated

constraints.

> Note that the use of constraints are optional but this can be useful in big scripts or multiple scripts where the type system could help you avoid bugs in your DataWeave logic by running the type checking algorithm.

Reviewer 2: `the use of constraints are optional` -> `the use of constraints is optional`

Author: Updated

The types supported by the DataWeave type system are divided into three categories:

* Simple Types
* Composite Types
* Complex Types

== Simple Types

> Simple types represent values such as String, Boolean, etc. These values are atomic, they are not composed by other values. The list of simple types are:

Reviewer 2: `The list of simple types are` -> `These are the simple types`

Author: Updated

> Simple types represent values such as String, Boolean, etc. These values are atomic, they are not composed by other values. These are the simple types:

Reviewer 3: ```suggestion
Simple types represent values such as strings, booleans, etc. These values are atomic, they are not composed by other values. These are the simple types:
```
The values are strings, booleans. The types are String, Boolean. Just a little detail that I missed.

* String
* Boolean
* Number
* Regex
* Null
* Temporal: Date, DateTime, LocalDateTime, LocalTime, Time, Period
//Key, KeyValuePairType, NameType, NameValuePairType, NamespaceType, RangeType , ReferenceType, TypeParameter, TypeType, UriType

=== Null

> Null is a type of only one value, which means that `null` can not be assigned to the type String or any other type except for the type Null.

Reviewer 2: `cannot`

> Null is a type of only one value, which means that `null` cannot be assigned to the type String or any other type except for the type Null.

Reviewer 3: ```suggestion
Null is a type of only one value, the value `null`, which means that this value `null` cannot be assigned to the type String or any other type except for the type Null.
```
Just to distinguish more between Null (that is the type) and null (that is the value).

> For example, the repeat function in the String Module has the signature `repeat(String, Number): String`, which means that only accepts as a first parameter a value of type String, and as a second parameter a value of type Number. The function returns a value of type String.

Reviewer 2: `it only accepts`

> ----

Reviewer 2: Is there a way to indicate that all these examples are DW code? I cannot remember.

Author: @dukesphere could you confirm?

repeat("a", 3)
----

This returns `&quot;aaa&quot;`, but:

----
repeat(null, 3)
----

> Throws an error because the value `null` can not be assigned to the parameter that expects a value of type String.

Reviewer 2: `cannot`

----
repeat(Null, 3)
----

In this case we are assigning to the first parameter a value that is the type Null (the type of this value

> is `Type&lt;Null&gt;`), and returns `&quot;NullNullNull&quot;` because there is an autocoercion that coerce the value to be of type String (`&quot;Null&quot;`).

Reviewer 2: `that coerces`

Some other functions may have an overloaded signature for the function (the function has multiples definitions for different parameters types) that let it be called with `null`.

> For example, the map function has the signature `map(Array&lt;T&gt;, (T, Number)-&gt;R): Array&lt;R&gt;` (receives an Array of elements of type `T` (this is a type parameter), a function, and returns an Array of elements of type `R`) but it is also defined as `map(Null, (Nothing, Nothing)-&gt;Any): Null`. +

Reviewer 2: @TomoDC are we covering type parameters here?

Reviewer 3: No, I chose to explain the map function because it is one of the most used functions and the signature results to have the type parameter :S. We could change to some other function without type params like isNumeric from the dw::core::String (there are a lot of functions there defined for null)

Reviewer 2: Maybe that would be best, cause otherwise you are forced to talk about a subject you haven't explained yet.

Author: @TomoDC Could you confirm the function to change and text?

Reviewer 3: ```suggestion
For example, the `isNumeric` function from dw::core::Strings has the signature `isNumeric(String): Boolean` (receives a string and returns if that string is numeric) but it is also defined as `isNumeric(Null): Boolean. +
```

Also change the next paragraph with:
`So if you call the function isNumeric with a null (isNumeric(null)) this does not throw an exception.
At runtime, the function dispatching algorithm selects the correct function based on the actual types of the values used. The function returns false because of the definition of isNumeric for null as the argument.`

Author: Updated.

So if you call the function map with a null (`null map (item, index) -&gt; item`) this does not throw an exception because map is defined for null as its first parameter. +

> The type system chooses the correct function definition with the function dispatching algorithm.

Reviewer 2: This sounds odd. I'd say something more like `At runtime, the function dispatching algorithm selects the correct function based on the actual types of the values used` @TomoDC @machaval please correct me if I'm wrong

Author: Updated.

This returns null because of the definition of map for null.

== Composite Types

Composite types contains other values inside it. These composite types are Array, Object and Function.

=== Array Type

An Array is of type `Array&lt;T&gt;` where `T` is a type parameter that defines the type of the elements inside the Array. +

> For example, the syntax to define an Array of Number as `Array&lt;Number&gt;` is:

Reviewer 3: ```suggestion
For example, the syntax to define a variable with the type Array of Number is:
```

----
var idsList: Array<Number> = [1, 22, 333, 4444]
----

The variable `idsList` only passes the type checking phase when the array assigned to it only contains Numbers.

=== Object Type

In order to define an object type there are two ways, either `Object` or `{}`, they both define an _open object_ and does not specify any constraint to their key value pairs. Open and close objects are later defined in this section.

The syntax to define an object type with a set of constraints for its key value pairs, is very similar to defining an object value:

----
{
    keyName @(attrName: AttrType): valueType,
    ...
}
----

For example, to define a user type with `firstName` String, `lastName` String and `age` Number, the syntax is:

----
type User = {
    firstName: String,
    lastName: String,
    age: Number
}
----

You can apply some modifiers to each key value pair to determine if a field is always present or if the field can be repeated or not.

For repeated fields use `*`. For example, the following type allows to have multiple `lastName` fields:

----
{
    firstName: String,
    lastName*: String,
    age: Number
}
----

> For conditional fields use `?`. For example, the following type allows the age field to be present or not:

Reviewer 2: @machaval did this make it into 2.3.0?

----
{
    firstName: String,
    lastName*: String,
    age?: Number
}
----

==== Close and Open Objects

Close object means that the type only accepts an object value if there are no additional fields in the object,
just those key values specified by the type.

> Open object only puts constraints on the fields that are declared in the type, and if they are all verified, then the type accepts the value. The type accept objects that have other fields apart of the ones mentioned in the declaration. All object types are open unless specified.

Reviewer 2: `Open objects only put`

The following example works even if `age` is not mentioned in the type declaration because the type of the variable `user` is an open object:

----
var user: {firstName: String, lastName: String} = {firstName: "John", lastName: "Smith", age: 34}
----

> The syntax to specify a close object is `{| |}`. The following example, makes our user type only supports the keys `firstName` and `lastName`, so this throws an exception because of the `age` field is not accepted.

Reviewer 2: `The following example, makes our user type only support the keys `firstName` and `lastName`, so this throws an exception because the `

----
var user: {|firstName: String, lastName: String|} = {name: "John", lastName: "Smith", age: 34}
----

=== Function Type

DataWeave is a functional language and functions are a first class citizen which means that functions are values
and values have an associated type.

The syntax to define a function type is:

----
(paramType: ParamType,...) -> ReturnType
----

> For example, if you want to define a function that has a parameter of type String, a second parameter of type Number and return a Boolean, the correct syntax is:

Reviewer 3: ```suggestion
For example, if you want to define in a constraint a function type that has a parameter of type String, a second parameter of type Number and return a Boolean, the correct syntax is:
```
I missed the difference between defining a function and defining a function type for a constraint. (little details haha but important)

----
(paramA: String, paramB: Number) -> Boolean
----

For example, you could define a function that receives another function as an argument:

----
fun findUserIndex(arr: Array<User>, condition: (userInArray: User, targetUser: User) -> Boolean): User = ???
----

> If we call the function `findUserIndex` with some function that does not match the `condition` constraint `(User, User) -> Boolean`, DataWeave throws a type checking error.

Reviewer 2: @TomoDC we could use examples here.

Reviewer 3: `fun applyIDsChange(ids: Array<Number>, changeTo: (Number) -> Number): Array<Number> = ???`

`applyIDsChange([1,2,3], (item: Number) -> item as String)`

Maybe this example is simpler than the findUserIndex. The function call throws an exception because the function argument is of type (Number) -> String instead of (Number) -> (Number).

Author: @afelisatti @TomoDC let me know if you would like to change the function and text with the new sample Tomo provided in his last comment.

Reviewer 3: Yep, I agree with adding the example.
```suggestion
If we call the function `applyIDsChange` with some function that does not match the `changeTo` constraint `(Number) -> Number, DataWeave throws a type checking error. 
For example in this function call `applyIDsChange([1,2,3], (item: Number) -> item as String)` DataWeave throws an exception because the function argument is of type `(Number) -> String` instead of `(Number) -> Number`.
```
Also change the findUserIndex with:
`fun applyIDsChange(ids: Array<Number>, changeTo: (Number) -> Number): Array<Number> = ???`

Author: Updated.

== Complex Types

=== Any and Nothing

> In some cases you can not enforce a restriction as it accepts all the values. +

Reviewer 2: `cannot`

Author: Updated

* `Any` type accepts all possible values.
* `Nothing` type accepts no value but it can be assigned to all the types. This type is not frequently used explicitly but it is by the type inference algorithm.

=== Union Type

Union type is used to compose types. The syntax to define a union type is:

----
TypeA | TypeB | ...
----

The following example defines a type that accepts any String and any Number:

----
var age: String | Number = ???
----

> A common patter is to use Null with a union type to specify that a type accepts null as a value. In the following example, the variable `a` accepts that the expression `payload.foo` returns `null`.

Reviewer 2: `pattern`

Author: Updated

> The following example defines a type that accepts any String and any Number:
> 
> ----
> var age: String | Number = ???
> ----

Reviewer 3: ```suggestion
The following example defines a variable with a constraint that accepts a value of type String or Number:

----
var age: String | Number = if (payload.allStrings) "32" else 32
----
```
This could be a good example

----
var a: String| Null = payload.foo
----

=== Intersection Type

The intersection type intersects object types. In this case, the intersection works as the concatenation (`++`) of object types.

_Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._

The syntax for intersection type is:
----
TypeA & TypeB & ...
----

In the following example, the intersection concatenates the two object types resulting in the type `{name: String, lastName: String}` This is an open object so it can have more key value pairs. The variable accepts the value assigned to it:

----
var a: {name: String} & {lastName: String} = {name: "John", lastName: "Smith", age: 34}
----

In the case of close objects, it returns the concatenation of the object types but resulting in a close object. +
In the following example, the intersection results in the type `{|name: String, lastName: String|}`. This throws
an exception because a close object does not accept an object value if there are additional fields in the object (field `age`):

----
var a: {|name: String|} & {|lastName: String|} = {name: "John", lastName: "Smith", age: 34}
----

=== Literal Types

> The literal type represents exactly one value. For example, the String value `"foo"` can be represented with the type `"foo"`.

Reviewer 2: `A literal type`?

Author: Updated

_Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._

The following literal types have been added to the type system:

* String literal types
* Number literal types
* Boolean literal types

You can use literal types with union types when you want to declare a type as a finite set of allowed values. For example, the following types declarations are aliases of union and literals types:

----
type Weekdays = "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday"

> type Days = "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" | "Saturday" | "Sunday"

Reviewer 2: @TomoDC could we do here `Weekdays | "Saturday" | "Sunday"`?

> You can use literal types with union types when you want to declare a type as a finite set of allowed values. For example, the following types declarations are aliases of union and literals types:
> 
> ----
> type Weekdays = "Saturday" | "Sunday"
> type Days = "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" | "Saturday" | "Sunday"

Reviewer 3: ```suggestion
You can use literal types with union types when you want to declare a type as a finite set of allowed values. For example, the following types declarations are aliases for unions of literal types:

----
type Weekdays = "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" 
type Days = Weekdays | "Saturday" | "Sunday"
```
Weekdays is used as type in the next type declaration (this is a good practice I think).

----

> In the following example, the variable only accepts the specific literal types (`404` and `500`). The type system ensures that the variable is one of those literal values:

Reviewer 3: ```suggestion
In the following example, the variable only accepts a value of these specific literal types (`404` and `500`). The type system ensures that the variable is one of those literal values:
```

----
var errorStatusCode: 404 | 500 = payload.statusCode
----

With function overloading you are able to define different behaviors based on the input argument value:

----
fun errorHandling(errorCode: 404 | 405, response): String = "Code for error 4XX handling here!"
fun errorHandling(errorCode: 500 | 501, response): String = "Code for error 5XX handling here!"
---
errorHandling(payload.statusCode, payload)
----

> The type system chooses the correct function definition depending on the value of `payload.statusCode`. It is not necessary to check the value of an argument with an if statement to change the behaviour of the function.

Reviewer 2: Should we reference the type system as the one making the choice? @TomoDC @machaval

Reviewer 3: ```suggestion
At runtime, the function dispatching algorithm selects the correct function based on the actual type of the value of `payload.statusCode`. It is not necessary to check the value of an argument with an if statement to change the behavior of the function.
```
What I wrote is probably wrong. I didn't notice that the function dispatching works at runtime so the TS didn't participate in the dispatching.
The proposed suggestion is more accurate.

Author: Updated.

= Type System
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

DataWeave 2.0 supports a type system. To take advantage of the type checking that the type system executes, you need to provide constraint expressions for variables and functions described in this section. For information on how values for these types are defined, refer to xref:dataweave-types.adoc[Value Constructs for Types].

A type system defines a set of constraints to a set of constructs such as:

* Variables
* Function parameters

These constraints are used in the type checking phase where DataWeave ensures that the values assigned
to variables or the arguments for a function call respects its constraints.

For example:

----
%dw 2.0
import * from dw::core::Strings
output application/json

var userName: String = "John"
----

The variable definition for `userName` has the constraint of assigning a value of type String.
If some other type value is assigned, a type checking error is raised.

----
%dw 2.0
import * from dw::core::Strings
output application/json

fun toUser(id: Number, userName: String): String = "you called the function toUser!"
----

> The constraints to call the function `toUser` are that the first argument has to be of type Number, the second one of type String (or to be coercible to those types). Another constraint here is that the result of the function has to be of type String.(this constraint isn't used in the call for `toUser`).

Reviewer: punctutation issue here: "Another constraint here is that the result of the function has to be of type String.(this constraint isn't used in the call for `toUser`).'

Reviewer 2: @kevintroller this issue is back, not sure why this is not updating.

Author: Seems Github does not reflect the change under the Conversation tab. However, the change is implemented correctly when you go to the Files Changed tab which is the latest version.

> The constraints to call the function `toUser` are that the first argument has to be of type Number, the second one of type String (or to be coercible to those types). Another constraint here is that the result of the function has to be of type String.(this constraint isn't used in the call for `toUser`).

Reviewer 2: `has to be of type String.(this` -> `has to be of type String (this`

Reviewer 2: @TomoDC should we say that the return type is used to check the function body? Something like `this constraint isn't used in the call for "toUser", it's applied to its definition to validate that the function body generates the proper type`

Author: Updated.

Reviewer 2: @kevintroller I don't see this updated. I see the punctuation thing and the old text lacking an explanation.

Author: Strange. I do see the change. If you navigate to the `Files Changed` tab in github does it reflect there?

----
toUser(123, "John")
toUser("123", true)
----

Though you would expect the second call to fail, autocoercions allow it to be successful and both of them work. In the case of the second function call, there is an autocoercion of the String `"123"` to the Number value `123`, and the Boolean `true` to the String value `"true"` to being able to call the function. +
An example of a call that throws a type checking error is `toUser("a 12", "John")` because the String `"a 12"` isn't coercible to a Number. +

DataWeave also uses a global type inference algorithm to validate your code even if no types are specified as
constraints.

> Note that the use of constraints is optional but this can be useful in big scripts or multiple scripts where the type system could help you avoid bugs in your DataWeave logic by running the type checking algorithm.

Reviewer 2: Maybe this should be an actual NOTE

The types supported by the DataWeave type system are divided into three categories:

* Simple Types
* Composite Types
* Complex Types

== Simple Types

Simple types represent values such as string, Booleans, etc. These values are atomic, they are not composed by other values. These are the simple types:

* String
* Boolean
* Number
* Regex
* Null
* Temporal: Date, DateTime, LocalDateTime, LocalTime, Time, Period
//Key, KeyValuePairType, NameType, NameValuePairType, NamespaceType, RangeType , ReferenceType, TypeParameter, TypeType, UriType

=== Null

Null is a type of only one value, the value `null`, which means that this value `null` cannot be assigned to the type String or any other type except for the type Null.

For example, the repeat function in the String Module has the signature `repeat(String, Number): String`, which means that it only accepts as a first parameter a value of type String, and as a second parameter a value of type Number. The function returns a value of type String.

----
%dw 2.0
import * from dw::core::Strings
output application/json

> 

Reviewer 2: Missing `---`

repeat("a", 3)
----

This returns `&quot;aaa&quot;`, but:

----
%dw 2.0
import * from dw::core::Strings
output application/json

> 

Reviewer 2: Same

repeat(null, 3)
----

Throws an error because the value `null` cannot be assigned to the parameter that expects a value of type String.

----
%dw 2.0
import * from dw::core::Strings
output application/json

> 

Reviewer 2: Same

Some other functions may have an overloaded signature for the function (the function has multiples definitions for different parameters types) that let it be called with `null`.

For example, the `isNumeric` function from dw::core::Strings has the signature `isNumeric(String): Boolean` (receives a string and returns if that string is numeric) but it is also defined as `isNumeric(Null): Boolean`. +

So if you call the function `isNumeric` with a null (`isNumeric(null)`) this does not throw an exception. At runtime, the function dispatching algorithm selects the correct function based on the actual types of the values used. The function returns false because of the definition of `isNumeric` for null as the argument.

== Composite Types

Composite types contains other values inside it. These composite types are Array, Object and Function.

=== Array Type

An Array is of type `Array&lt;T&gt;` where `T` is a type parameter that defines the type of the elements inside the Array. +
For example, the syntax to define a variable with the type Array of `Number` is:

----
%dw 2.0

> import * from dw::core::Arrays

Reviewer 2: This is not used on this script

output application/json

var idsList: Array<Number> = [1, 22, 333, 4444]
----

The variable `idsList` only passes the type checking phase when the array assigned to it only contains Numbers.

=== Object Type

> In order to define an object type there are two ways, either `Object` or `{}`, they both define an open object and doesn't specify any constraint to their key value pairs. Open and close objects are later defined in this section.

Reviewer 2: `doesn't` -> `don't`?

The syntax to define an object type with a set of constraints for its key value pairs, is very similar to defining an object value:

----
{
    keyName @(attrName: AttrType): valueType,
    ...
}
----

For example, to define a user type with `firstName` String, `lastName` String and `age` Number, the syntax is:

----
%dw 2.0
output application/json

type User = {
    firstName: String,
    lastName: String,
    age: Number
}
----

You can apply some modifiers to each key value pair to determine if a field is always present or if the field can be repeated or not.

For repeated fields use `*`. For example, the following type allows to have multiple `lastName` fields:

----
%dw 2.0
output application/json

{
    firstName: String,
    lastName*: String,
    age: Number
}
----

For conditional fields use `?`. For example, the following type allows the age field to be present or not:

----
%dw 2.0
output application/json

{
    firstName: String,
    lastName*: String,
    age?: Number
}
----

==== Close and Open Objects

Close object means that the type only accepts an object value if there are no additional fields in the object,
just those key values specified by the type.

> Open object only put constraints on the fields that are declared in the type, and if they are all verified, then the type accepts the value. The type accept objects that have other fields apart of the ones mentioned in the declaration. All object types are open unless specified.

Reviewer 2: `The type accept` -> `The type accepts`

----
%dw 2.0
output application/json

var user: {|firstName: String, lastName: String|} = {name: "John", lastName: "Smith", age: 34}
----

=== Function Type

DataWeave is a functional language and functions are a first class citizen which means that functions are values
and values have an associated type.

The syntax to define a function type is:

----
(paramType: ParamType,...) -> ReturnType
----

For example, if you want to define in a constraint a function type that has a parameter of type String, a second parameter of type Number and return a Boolean, the correct syntax is:

----
(paramA: String, paramB: Number) -> Boolean
----

For example, you could define a function that receives another function as an argument:

----
%dw 2.0
output application/json

fun applyIDsChange(ids: Array<Number>, changeTo: (Number) -> Number): Array<Number> = ???
----

If we call the function `applyIDsChange` with some function that does not match the `changeTo` constraint `(Number) -> Number`, DataWeave throws a type checking error.

> 

Reviewer 2: We are missing an example here.

> 

Reviewer 2: ```suggestion
For example, the following call works because `abs` is a function that takes a Number and returns a Number, which matches the `changeTo` parameter.

----
applyIDsChange([1,-6, 3, -8], abs)
----

But this call fails because `sum` is a function that takes an Array and returns a Number, so it does not match our parameter constraint.

----
applyIDsChange([1,-6, 3, -8], sum)
----
```

== Complex Types

=== Any and Nothing

In some cases you can not enforce a restriction as it accepts all the values. +

* `Any` type accepts all possible values.
* `Nothing` type accepts no value but it can be assigned to all the types. This type is not frequently used explicitly but it is by the type inference algorithm.

=== Union Type

Union type is used to compose types. The syntax to define a union type is:

----
TypeA | TypeB | ...
----

The following example defines a variable with a constraint that accepts a value of type String or Number:

----
%dw 2.0

> import * from dw::core::Strings

Reviewer 2: This is not used here.

output application/json

var age: String | Number = if (payload.allStrings) "32" else 32
----

A common pattern is to use Null with a union type to specify that a type accepts null as a value. In the following example, the variable `a` accepts that the expression `payload.foo` returns `null`.

----
%dw 2.0

> import * from dw::core::Strings

Reviewer 2: Same

output application/json

var a: String| Null = payload.foo
----

=== Intersection Type

The intersection type intersects object types. In this case, the intersection works as the concatenation (`++`) of object types.

_Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._

The syntax for intersection type is:
----
TypeA & TypeB & ...
----

In the following example, the intersection concatenates the two object types resulting in the type `{name: String, lastName: String}` This is an open object so it can have more key value pairs. The variable accepts the value assigned to it:

----
%dw 2.0

> import * from dw::core::Strings

Reviewer 2: Same

output application/json

var a: {name: String} & {lastName: String} = {name: "John", lastName: "Smith", age: 34}
----

In the case of close objects, it returns the concatenation of the object types but resulting in a close object. +
In the following example, the intersection results in the type `{|name: String, lastName: String|}`. This throws
an exception because a close object does not accept an object value if there are additional fields in the object (field `age`):

----
%dw 2.0

> import * from dw::core::Strings

Reviewer 2: Same
